### ISSUE TITLE
transaction leaves inconsistent result

### ISSUE THREAD

**Issue by matthiasclasen** [flatpak/flatpak#2196]
I was writing tests for FlatpakTransaction, and I wrote one that started with a repo containing org.test.Platform, and then did:

> flatpak_transaction_add_install ( ... org.test.Hello ...)
> flatpak_transaction_add_uninstall (... org.test.Platform ...)
> flatpak_transaction_run (...)

I was expecting to get an error about not being able to uninstall the runtime because the app depends on it, but the transaction happily goes ahead, and produces a repo that only contains a single ref,
org.test.Hello.

I then added another transaction, that did:

> flatpak_transaction_add_update (... org.test.Hello ...)
> flatpak_transaction_run (...)

This one surprised me by bringing org.test.Platform back, despite the update itself being a no-op.

In the end, I am confused: Are transactions supposed to guarantee dependency closure, or not?

**Comment by matthiasclasen** [comment]
I think what might be happening is that we add the uninstall, which is fine, then we add the install, which resolves fine without adding a runtime install because it sees the runtime is available.
Then when we execute, we uninstall the runtime just fine - there's no app needing it yet.
And later on we install the app :(

**Comment by alexlarsson** [comment]
Yeah, thats whats happening. I guess we need to take uninstalls into consideration when seeing what is available.

I'm not sure what the expected result should be though. Uninstall + reinstall, or just remove the uninstall op?

**Comment by matthiasclasen** [comment]
I was expecting an error, tbh. Unless force-delete is set
